Update designCarousel to include link-in-bio related changes.

agrullon95 opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #76719

## Proposed Changes

* DesignCarousel and Pattern-picker packages have similar functionality, so it is best to try to consolidate them for easier maintainability. This PR updates the designCarousel package to work with link-in-bio flow. 

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Checkout this branch
* Execute `yarn start`
* Navigate to http://calypso.localhost:3000/setup/ecommerce/designCarousel to ensure it is working as intended and looks correct.
* Navigate to http://calypso.localhost:3000/setup/link-in-bio/patterns to ensure it is working as intended and looks correct.
* Navigate to http://calypso.localhost:3000/setup/link-in-bio-tld/patterns to ensure it is working as intended and looks correct.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

agrullon95 committed:
Update designCarousel to include link-in-bio related changes.

agrullon95 committed:
Update design-picker test cases

agrullon95 committed:
Update getDesignPreivewUrl site_title

agrullon95 commented:
The Link-in-bio mobile viewport still needs to be tweaked to match production:

![CleanShot 2023-05-17 at 00 19 03@2x](https://github.com/Automattic/wp-calypso/assets/10482592/ea3d12ab-1814-42e0-a0d3-311d359d8116)

![CleanShot 2023-05-17 at 00 19 30@2x](https://github.com/Automattic/wp-calypso/assets/10482592/e0dbf924-f855-4949-a1e4-301b8e828fe7)

agrullon95 committed:
Update dependencies

escapemanuele commented on client/landing/stepper/declarative-flow/link-in-bio.ts:
```diff
@@ -31,7 +31,11 @@ const linkInBio: Flow = {
 			{ slug: 'domains', asyncComponent: () => import( './internals/steps-repository/domains' ) },
 			{ slug: 'plans', asyncComponent: () => import( './internals/steps-repository/plans' ) },
-			{ slug: 'patterns', asyncComponent: () => import( './internals/steps-repository/patterns' ) },
+
```
We don't need the empty line here 👍

escapemanuele commented on packages/design-carousel/package.json:
```diff
@@ -31,6 +31,7 @@
 	"dependencies": {
 		"@automattic/i18n-utils": "workspace:^",
 		"@automattic/onboarding": "workspace:^",
+		"@automattic/design-picker": "workspace:^",
```
I think you may need to run `yarn` in `packages/design-carousel` and commit the `yarn.lock` and `package.json` files 👍

agrullon95 replied:
Thanks. I also added it to the tsconfig.json file. Now its working.

escapemanuele commented:
Hey @agrullon95, great PR! I tested and it is working, I left a couple of comments to address and then maybe the tests will pass 🚀

agrullon95 committed:
Remove empty space

agrullon95 committed:
Update package.json for design-carousel

agrullon95 committed:
Update tsconfig to reference design-picker

agrullon95 committed:
Remove unneccesary component prop

agrullon95 committed:
Update designCarousel import for other lib related flows

agrullon95 committed:
Revert pattern picker imported update

agrullon95 committed:
Revert pattern picker flow prop

simison commented on client/landing/stepper/declarative-flow/internals/steps-repository/design-carousel/style.scss:
```diff
@@ -1,5 +1,123 @@
+}
+
+.link-in-bio.patterns,
+.link-in-bio-tld.patterns {
```
Should these be just "mobile patterns" instead of scoping them to a single flow? Makes it less cumbersome to re-use for other flows in future.

agrullon95 replied:
Yes thats a great idea. I can make these changes in a follow up PR. That should make it easier to add flows without worrying about necessary CSS changes.

simison replied:
Yes, totes fine for a follow-up. Thank you!

Remember to consider re-usability across all code, CSS, JS, etc.

daledupreez commented on packages/design-carousel/src/index.tsx:
```diff
@@ -1,9 +1,12 @@
+import { Gridicon } from '@automattic/components';
+import { StarterDesigns, useStarterDesignsQuery } from '@automattic/data-stores';
 import { useLocale } from '@automattic/i18n-utils';
+import { MShotsOptions } from '@automattic/onboarding';
```
Should this not be a type import?
```suggestion
import type { MShotsOptions } from '@automattic/onboarding';
```

daledupreez commented on packages/design-carousel/src/index.tsx:
```diff
@@ -101,6 +192,8 @@ export default function DesignCarousel( { onPick, selectedDesignSlugs }: DesignC
 					} }
 				>
 					<span>{ __( 'Continue' ) }</span>
+					<Gridicon icon="heart" size={ 18 } />
+					{ /* Heart icon when hovering over continue missing here */ }
```
What is this comment for? I am a bit unclear on the intent.

agrullon95 replied:
This was a leftover comment for me. I removed it.

daledupreez commented on packages/design-picker/src/utils/designs.ts:
```diff
@@ -33,7 +33,7 @@ export const getDesignPreviewUrl = (
 			: undefined,
 		source_site: 'patternboilerplates.wordpress.com',
-		use_screenshot_overrides: options.use_screenshot_overrides,
+		use_screenshot_overrides: options.use_screenshot_overrides || true,
```
This feels like it should either be `?? true` or just `true`, as the `||` will fall through to `true` if the incoming value is falsey.

agrullon95 replied:
@daledupreez  Good catch.

What do you think about changing it to:

```
...( options.use_screenshot_overrides && {
			use_screenshot_overrides: options.use_screenshot_overrides,
		} ),
```

This way its added only when necessary and I can update the design-carousel item component:

```
<MShotsImage
	url={ getDesignPreviewUrl( design, {
	     use_screenshot_overrides: true,
	} ) }
	options={ options }
	alt={ design.title }
        aria-labelledby=""
/>
```

daledupreez replied:
I think that should work!

daledupreez reviewed (commented):
I have some comments up from inspection, which likely need a bit of discussion.

agrullon95 committed:
Remove leftover comment

agrullon95 committed:
Update MshotsOption to type import

alshakero commented on packages/design-carousel/src/index.tsx:
```diff
@@ -1,45 +1,121 @@
-	let selectedDesigns = allDesigns?.designs;
+	const shouldDisplayMobileViewOnly = ( flow: string | undefined ) => {
+		switch ( flow ) {
+			case 'link-in-bio':
```
Let's use [`isLinkInBioFlow`](https://github.com/Automattic/wp-calypso/blob/trunk/packages/onboarding/src/utils/flows.ts#LL28C14-L28C29) instead. And since this function doesn't depend on the state, we can hoist it up and outside the component; to avoid recreating it at every render.

alshakero commented on packages/design-carousel/src/index.tsx:
```diff
@@ -1,45 +1,121 @@
 	} );
 
-	let selectedDesigns = allDesigns?.designs;
+	const shouldDisplayMobileViewOnly = ( flow: string | undefined ) => {
+		switch ( flow ) {
+			case 'link-in-bio':
+				return true;
+			default:
+				return false;
+		}
+	};
```
```suggestion
	const shouldDisplayMobileViewOnly = ( flow: string | undefined ) => {
		switch ( true ) {
			case isLinkInBioFlow(flow):
				return true;
			default:
				return false;
		}
	};
```

alshakero commented on packages/design-carousel/src/index.tsx:
```diff
@@ -1,45 +1,121 @@
-			selectedDesignSlugs.includes( design.slug )
-		);
+	const getLinkInBioDesigns = ( allDesigns: StarterDesigns | undefined ) => {
+		const designs = allDesigns
+			? allDesigns?.designs.filter(
+					( design ) =>
+						design.is_virtual &&
+						design.categories.some( ( category ) => category.slug === 'link-in-bio' )
+			  )
+			: [];
```
```suggestion
		const designs = allDesigns?.designs.filter(
					( design ) =>
						design.is_virtual &&
						design.categories.some( ( category ) => category.slug === 'link-in-bio' )
			  )
			?? [];
```

alshakero commented on packages/design-carousel/src/index.tsx:
```diff
@@ -1,45 +1,121 @@
+			return null;
+		}
+
+		switch ( flow ) {
+			case 'link-in-bio':
```
Same as above, this may break for other variations of LiB flow (eg `link-in-bio-tld`).

agrullon95 replied:
Good catch. I updated these cases to use the `isLinkInBioFlow` function.

alshakero commented on packages/design-carousel/src/index.tsx:
```diff
@@ -1,45 +1,121 @@
+		switch ( flow ) {
+			case 'link-in-bio':
+				return getLinkInBioDesigns( allDesigns );
+			case 'ecommerce':
```
Let's use [constants](https://github.com/Automattic/wp-calypso/blob/2693b31ceeb00c7cf352c69255311641280637ed/packages/onboarding/src/utils/flows.ts#L13).

agrullon95 replied:
Updated to use constants for ecommerce and link-in-bio.

alshakero commented on packages/design-carousel/src/index.tsx:
```diff
@@ -1,45 +1,121 @@
-		const filteredDesigns = selectedDesigns.filter( ( design ) =>
-			selectedDesignSlugs.includes( design.slug )
-		);
+	const getLinkInBioDesigns = ( allDesigns: StarterDesigns | undefined ) => {
```
Should be outside the component.

alshakero commented on packages/design-carousel/src/index.tsx:
```diff
@@ -1,45 +1,121 @@
+		return designs;
+	};
+
+	const getEcommerceDesigns = (
```
Should be outside the component. But would need `selectedDesign` argument.

agrullon95 replied:
I moved this function outside. Im not sure if I need the selectedDesign argument since its derived from allDesigns.

alshakero commented on packages/design-carousel/src/index.tsx:
```diff
@@ -1,45 +1,121 @@
+
+	const getCarouselOptions = ( flow: string | undefined ) => {
+		switch ( flow ) {
+			case 'link-in-bio':
```
Should use `isLinkInBioFlow`.

alshakero commented on packages/design-carousel/src/index.tsx:
```diff
@@ -1,45 +1,121 @@
+		return selectedDesigns;
+	};
+
+	const getCarouselOptions = ( flow: string | undefined ) => {
```
Can be outside the component.

alshakero commented on packages/design-carousel/src/index.tsx:
```diff
@@ -1,45 +1,121 @@
+		}
+	};
+
+	const getFlowDesigns = (
+		allDesigns: StarterDesigns | undefined,
+		locale: string | undefined,
+		selectedDesignSlugs: string[] | undefined,
+		flow: string | undefined
+	) => {
```
Should be outside the component.

alshakero reviewed (changes requested):
Left mostly optional feedback. Really good work. Should be tested with LiB flow variations because I think it's currently broken in two of them.

agrullon95 committed:
Move functions outside of component

agrullon95 commented:
> Left mostly optional feedback. Really good work. Should be tested with LiB flow variations because I think it's currently broken in two of them.

@alshakero Great feedback. I updated the PR to include the changes you suggested above.
Including using constants for the flows, using the `isLinkInBioFlow` function to cover the different link-in-bio flows, and moved the functions outside of the components.

agrullon95 committed:
Add use_screenshot_overrides to getDesignPreviewUrl function.

agrullon95 committed:
Use use_screenshot_override property in design-carousel Item component.

agrullon95 committed:
Revert test case changes related to use_screenshot_overrides

daledupreez commented on packages/design-carousel/src/index.tsx:
```diff
@@ -12,19 +15,33 @@ import type { Design } from '@automattic/design-picker/src/types';
 type DesignCarouselProps = {
 	onPick: ( design: Design ) => void;
 	selectedDesignSlugs?: string[];
+	flow?: string | null | undefined;
```
I have a general comment here: can we not move _all_ the logic from functions below into props that are specified by callers? This should simplify the component and avoid reaching back out into external logic. As things stand, there is a _lot_ of flow-specific logic here, and it feels like pushing some of the decision-making up to the calling component will be easier to maintain in future.
Essentially, I am thinking we could update the props as per the suggestion here.
```suggestion
	flow?: string | null | undefined;
	designFilter?: ( design: Design ) => boolean; // My syntax could be wrong here; this would also mean `selectedDesignSlugs` could probably be dropped
	displayOnlyMobileView?: boolean;
	carouselDesktopOptions?: MShotsOptions; // should be defaulted if not present
	carouselMobileOptions?: MShotsOptions; // should be defaulted if not present
```

Maybe we can tackle that in a follow-up PR, but I think it would be a much cleaner interface.

agrullon95 replied:
@daledupreez - That makes sense and is definitely something I can change in a follow-up PR. If everything is working with the current PR on your end, I can merge this and start working on those changes.

alshakero reviewed (approved):
Such a neat PR now. Absolutely beautiful. 

Tested LiB and LiB TLD, and ecommerce.

daledupreez reviewed (approved):
Let's get this shipped!

I tested the eCommerce flow, and that looks good to me.
